client/client_test.go

```diff
@@ -6253,6 +6257,169 @@ func testExportAnnotations(t *testing.T, sb integration.Sandbox) {
+	}
+
+	p := platforms.MustParse("linux/amd64")
+	pk := platforms.Format(p)
+
+	success := []byte(`{"success": true}`)
+	successDigest := digest.SHA256.FromBytes(success)
+
+	frontend := func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+		res := gateway.NewResult()
+
+		// build image
+		st := llb.Scratch().File(
+			llb.Mkfile("/greeting", 0600, []byte("hello world!")),
+		)
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, err
+		}
+		r, err := c.Solve(ctx, gateway.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, err
+		}
+		ref, err := r.SingleRef()
+		if err != nil {
+			return nil, err
+		}
+		_, err = ref.ToState()
+		if err != nil {
+			return nil, err
+		}
+		res.AddRef(pk, ref)
+
+		// build attestations
+		st = llb.Scratch().
+			File(llb.Mkfile("/attestation.json", 0600, success)).
+			File(llb.Mkfile("/attestation2.json", 0600, []byte{}))
+		def, err = st.Marshal(ctx)
+		if err != nil {
+			return nil, err
+		}
+		r, err = c.Solve(ctx, gateway.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, err
+		}
+		refAttest, err := r.SingleRef()
+		if err != nil {
+			return nil, err
+		}
+		_, err = ref.ToState()
+		if err != nil {
+			return nil, err
+		}
+
+		expPlatforms := &exptypes.Platforms{
+			Platforms: []exptypes.Platform{{ID: pk, Platform: p}},
+		}
+		dt, err := json.Marshal(expPlatforms)
+		if err != nil {
+			return nil, err
+		}
+		res.AddMeta(exptypes.ExporterPlatformsKey, dt)
+
+		res.AddAttestation(pk, &gateway.InTotoAttestation{
+			PredicateRef:  refAttest,
+			PredicatePath: "/attestation.json",
+			PredicateType: "https://example.com/attestations/v1.0",
+			Subjects: []attestation.InTotoSubject{
+				&attestation.InTotoSubjectSelf{},
+			},
+		})
+		res.AddAttestation(pk, &gateway.InTotoAttestation{
+			PredicateRef:  refAttest,
+			PredicatePath: "/attestation2.json",
+			PredicateType: "https://example.com/attestations2/v1.0",
+			Subjects: []attestation.InTotoSubject{
+				&attestation.InTotoSubjectRaw{
+					Name:   "/attestation.json",
+					Digest: []digest.Digest{successDigest},
+				},
+			},
+		})
+
+		return res, nil
+	}
+
+	target := registry + "/buildkit/testattestations:latest"
+	_, err = c.Build(sb.Context(), SolveOpt{
+		Exports: []ExportEntry{
+			{
+				Type: ExporterImage,
+				Attrs: map[string]string{
+					"name": target,
+					"push": "true",
+				},
+			},
+		},
+	}, "", frontend, nil)
+	require.NoError(t, err)
+
+	desc, provider, err := contentutil.ProviderFromRef(target)
+	require.NoError(t, err)
+
+	index, err := testutil.ReadIndex(sb.Context(), provider, desc)
+	require.NoError(t, err)
+	require.Equal(t, 2, len(index))
+
+	img := index.Find(pk)
+	require.NotNil(t, img)
+	require.Equal(t, pk, platforms.Format(*img.Desc.Platform))
+	require.Equal(t, 1, len(img.Layers))
+	require.Equal(t, []byte("hello world!"), img.Layers[0]["greeting"].Data)
+
+	atts := index.Filter("unknown/unknown")
```

Comment:
This needs to check that the annotations pointing to the correct descriptor are correct. Ideally we would create two descriptors in here so we can check that each attestation manifest points to the correct descriptor.